### PR TITLE
Gm wfpr

### PR DESCRIPTION
**Pull request by mgalbier**
- Remove change indicator for move and modification, use connecting arrow for them
- Initial version of ObjectPathCrumb
- Improve if structure
  <a href='#crh-start'></a><a href='#crh-data-%7B%22processed%22%3A%20%5B%22https%3A//github.com/dimitar-asenov/Envision/pull/395%23discussion_r66703862%22%2C%20%22https%3A//github.com/dimitar-asenov/Envision/pull/395%23discussion_r66703866%22%2C%20%22https%3A//github.com/dimitar-asenov/Envision/pull/395%23discussion_r66703879%22%2C%20%22https%3A//github.com/dimitar-asenov/Envision/pull/395%23issuecomment-225348100%22%2C%20%22https%3A//github.com/dimitar-asenov/Envision/pull/395%23discussion_r66704020%22%2C%20%22https%3A//github.com/dimitar-asenov/Envision/pull/395%23discussion_r66704025%22%2C%20%22https%3A//github.com/dimitar-asenov/Envision/pull/395%23discussion_r66704029%22%2C%20%22https%3A//github.com/dimitar-asenov/Envision/pull/395%23discussion_r66704064%22%2C%20%22https%3A//github.com/dimitar-asenov/Envision/pull/395%23discussion_r66704101%22%2C%20%22https%3A//github.com/dimitar-asenov/Envision/pull/395%23issuecomment-225349402%22%2C%20%22https%3A//github.com/dimitar-asenov/Envision/pull/395%23discussion_r66760458%22%2C%20%22https%3A//github.com/dimitar-asenov/Envision/pull/395%23discussion_r66762672%22%2C%20%22https%3A//github.com/dimitar-asenov/Envision/pull/395%23discussion_r66763408%22%2C%20%22https%3A//github.com/dimitar-asenov/Envision/pull/395%23issuecomment-225532803%22%2C%20%22https%3A//github.com/dimitar-asenov/Envision/pull/395%23issuecomment-225534772%22%5D%2C%20%22comments%22%3A%20%7B%22General%20Comment%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov/Envision/pull/395%23issuecomment-225348100%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22I%20wouldn%27t%20delete%20the%20svgs%20and%20the%20styles%20corresponding%20to%20the%20yellow%20and%20blue%20arrows.%20Just%20leave%20them%20unused.%20Who%20knows%20if%20we%20change%20our%20minds%20in%20the%20future.%22%2C%20%22created_at%22%3A%20%222016-06-11T09%3A43%3A43Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/1152976%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov%22%7D%7D%2C%20%7B%22body%22%3A%20%22OK%20I%27m%20done%20with%20this%20review.%22%2C%20%22created_at%22%3A%20%222016-06-11T10%3A06%3A29Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/1152976%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov%22%7D%7D%2C%20%7B%22body%22%3A%20%22Thanks.%20Updated%22%2C%20%22created_at%22%3A%20%222016-06-13T09%3A33%3A32Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/17177211%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/mgalbier%22%7D%7D%2C%20%7B%22body%22%3A%20%22Nice%2C%20thanks.%22%2C%20%22created_at%22%3A%20%222016-06-13T09%3A42%3A17Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/1152976%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov%22%7D%7D%5D%2C%20%22title%22%3A%20%22General%20Comment%22%7D%2C%20%22Pull%20c3b565178724660cea8671850c8d12819a7ba4ef%20VersionControlUI/src/nodes/DiffComparisonPair.h%2018%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov/Envision/pull/395%23discussion_r66704064%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22You%20are%20storing%20pointers%20here%20and%20you%20did%20not%20explicitly%20delete%20those%20in%20the%20destructor%20so%20there%20is%20a%20memory%20leak.%5Cr%5Cn%5Cr%5CnUnless%20you%20have%20a%20very%20specific%20reason%20why%20you%20need%20pointers%20%28I%27d%20like%20hear%20it%20if%20so%29%2C%20store%20directly%20the%20structures%20here%20and%20avoid%20the%20hassle%20of%20dealing%20with%20memory.%22%2C%20%22created_at%22%3A%20%222016-06-11T10%3A01%3A11Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/1152976%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20VersionControlUI/src/nodes/DiffComparisonPair.h%3AL68-77%22%7D%2C%20%22Pull%20296d908ab525c1d933779a9420849d04f005024b%20VersionControlUI/src/DiffManager.cpp%2019%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov/Envision/pull/395%23discussion_r66703862%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22instead%20of%20assigning%20%5C%22%5C%22%20to%20the%20strings%20here%2C%20just%20leave%20them%20uninitialized.%20Below%20you%20can%20use%20%60isNull%28%29%60%20to%20check%20if%20the%20strings%20have%20changed.%22%2C%20%22created_at%22%3A%20%222016-06-11T09%3A40%3A11Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/1152976%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20VersionControlUI/src/DiffManager.cpp%3AL276-295%22%7D%2C%20%22Pull%20c3b565178724660cea8671850c8d12819a7ba4ef%20VersionControlUI/src/items/VDiffComparisonPair.cpp%20113%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov/Envision/pull/395%23discussion_r66704029%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22no%20clone%20on%20the%20two%20lines%20below%22%2C%20%22created_at%22%3A%20%222016-06-11T09%3A57%3A39Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/1152976%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20VersionControlUI/src/items/VDiffComparisonPair.cpp%3AL200-213%22%7D%2C%20%22Pull%20296d908ab525c1d933779a9420849d04f005024b%20VersionControlUI/src/DiffManager.cpp%2065%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov/Envision/pull/395%23discussion_r66703866%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22use%20%60isNull%28%29%60%22%2C%20%22created_at%22%3A%20%222016-06-11T09%3A40%3A25Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/1152976%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20VersionControlUI/src/DiffManager.cpp%3AL385-397%22%7D%2C%20%22Pull%20c3b565178724660cea8671850c8d12819a7ba4ef%20VersionControlUI/src/nodes/DiffComparisonPair.cpp%2087%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov/Envision/pull/395%23discussion_r66704101%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22I%27m%20a%20bit%20confused.%20Now%20that%20we%20have%20the%20bread%20crumbs%2C%20do%20we%20still%20need%20this%20component%20name%3F%20Maybe%20I%20just%20forgot%20what%20it%20is.%22%2C%20%22created_at%22%3A%20%222016-06-11T10%3A05%3A44Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/1152976%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov%22%7D%7D%2C%20%7B%22body%22%3A%20%22It%20is%20used%20for%20the%20sorting%20of%20the%20DiffComparisonPairs.%20But%20there%20are%20fields%20which%20are%20no%20longer%20needed.%20I%20remove%20them%20as%20well%20as%20the%20%5C%22%20%3E%20%5C%22%20at%20the%20beginning%20of%20componentName%2C%20which%20does%20not%20serve%20any%20purpose.%22%2C%20%22created_at%22%3A%20%222016-06-13T09%3A32%3A37Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/17177211%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/mgalbier%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20VersionControlUI/src/nodes/DiffComparisonPair.cpp%3AL155-176%22%7D%2C%20%22Pull%20296d908ab525c1d933779a9420849d04f005024b%20VersionControlUI/src/DiffManager.cpp%2044%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov/Envision/pull/395%23discussion_r66703879%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22why%20did%20you%20change%20this%20condition%3F%22%2C%20%22created_at%22%3A%20%222016-06-11T09%3A42%3A15Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/1152976%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov%22%7D%7D%2C%20%7B%22body%22%3A%20%22If%20one%20of%20the%20items%20is%20null%20it%20does%20not%20make%20sense%20to%20add%20an%20arrow%20between%20them.%20The%20if%20structure%20was%20changed%20in%20a%20later%20commit%20a%20bit.%20An%20outer%20if%20checks%20if%20the%20items%20are%20not%20null%20and%20only%20then%20checks%20the%20other%20stuff.%22%2C%20%22created_at%22%3A%20%222016-06-13T09%3A10%3A30Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/17177211%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/mgalbier%22%7D%7D%2C%20%7B%22body%22%3A%20%22OK%2C%20thanks%20for%20explaining.%22%2C%20%22created_at%22%3A%20%222016-06-13T09%3A27%3A17Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/1152976%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20VersionControlUI/src/DiffManager.cpp%3AL330-346%22%7D%2C%20%22Pull%20c3b565178724660cea8671850c8d12819a7ba4ef%20VersionControlUI/src/items/VDiffComparisonPair.cpp%2089%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov/Envision/pull/395%23discussion_r66704025%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22no%20clone%20on%20next%20line%20and%20on%20the%20%60addForm%60%20line%20following%22%2C%20%22created_at%22%3A%20%222016-06-11T09%3A57%3A21Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/1152976%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20VersionControlUI/src/items/VDiffComparisonPair.cpp%3AL160-182%22%7D%2C%20%22Pull%20c3b565178724660cea8671850c8d12819a7ba4ef%20VersionControlUI/src/items/VDiffComparisonPair.cpp%2064%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov/Envision/pull/395%23discussion_r66704020%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22In%20the%20line%20below%20and%20the%20%60addForm%60%20line%20you%20are%20unnecessarily%20cloning%20and%20thereby%20leaking%20memory.%20Don%27t%20clone%20anything%2C%20since%20you%20overwrite%20the%20%60infoGrid%60%20and%20%60diffGrid%60%20and%20%60container%60%20variables%20with%20new%20objects%20anyway%20later.%22%2C%20%22created_at%22%3A%20%222016-06-11T09%3A56%3A53Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/1152976%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20VersionControlUI/src/items/VDiffComparisonPair.cpp%3AL117-140%22%7D%7D%7D'></a>
  <a href='https://www.codereviewhub.com/'></a>
- [ ] <a href='#crh-comment-Pull 296d908ab525c1d933779a9420849d04f005024b VersionControlUI/src/DiffManager.cpp 19'></a> &nbsp;<b><a href='https://github.com/dimitar-asenov/Envision/pull/395#discussion_r66703862'>File: VersionControlUI/src/DiffManager.cpp:L276-295</a></b>
- <a href='https://github.com/dimitar-asenov'></a> instead of assigning "" to the strings here, just leave them uninitialized. Below you can use `isNull()` to check if the strings have changed.
- [ ] <a href='#crh-comment-Pull 296d908ab525c1d933779a9420849d04f005024b VersionControlUI/src/DiffManager.cpp 65'></a> &nbsp;<b><a href='https://github.com/dimitar-asenov/Envision/pull/395#discussion_r66703866'>File: VersionControlUI/src/DiffManager.cpp:L385-397</a></b>
- <a href='https://github.com/dimitar-asenov'></a> use `isNull()`
- [ ] <a href='#crh-comment-Pull 296d908ab525c1d933779a9420849d04f005024b VersionControlUI/src/DiffManager.cpp 44'></a> &nbsp;<b><a href='https://github.com/dimitar-asenov/Envision/pull/395#discussion_r66703879'>File: VersionControlUI/src/DiffManager.cpp:L330-346</a></b>
- <a href='https://github.com/dimitar-asenov'></a> why did you change this condition?
- <a href='https://github.com/mgalbier'></a> If one of the items is null it does not make sense to add an arrow between them. The if structure was changed in a later commit a bit. An outer if checks if the items are not null and only then checks the other stuff.
- <a href='https://github.com/dimitar-asenov'></a> OK, thanks for explaining.
- [ ] <a href='#crh-comment-General Comment'></a> &nbsp;<b><a href='https://github.com/dimitar-asenov/Envision/pull/395#issuecomment-225348100'>General Comment</a></b>
- <a href='https://github.com/dimitar-asenov'></a> I wouldn't delete the svgs and the styles corresponding to the yellow and blue arrows. Just leave them unused. Who knows if we change our minds in the future.
- <a href='https://github.com/dimitar-asenov'></a> OK I'm done with this review.
- <a href='https://github.com/mgalbier'></a> Thanks. Updated
- <a href='https://github.com/dimitar-asenov'></a> Nice, thanks.
- [ ] <a href='#crh-comment-Pull c3b565178724660cea8671850c8d12819a7ba4ef VersionControlUI/src/items/VDiffComparisonPair.cpp 64'></a> &nbsp;<b><a href='https://github.com/dimitar-asenov/Envision/pull/395#discussion_r66704020'>File: VersionControlUI/src/items/VDiffComparisonPair.cpp:L117-140</a></b>
- <a href='https://github.com/dimitar-asenov'></a> In the line below and the `addForm` line you are unnecessarily cloning and thereby leaking memory. Don't clone anything, since you overwrite the `infoGrid` and `diffGrid` and `container` variables with new objects anyway later.
- [ ] <a href='#crh-comment-Pull c3b565178724660cea8671850c8d12819a7ba4ef VersionControlUI/src/items/VDiffComparisonPair.cpp 89'></a> &nbsp;<b><a href='https://github.com/dimitar-asenov/Envision/pull/395#discussion_r66704025'>File: VersionControlUI/src/items/VDiffComparisonPair.cpp:L160-182</a></b>
- <a href='https://github.com/dimitar-asenov'></a> no clone on next line and on the `addForm` line following
- [ ] <a href='#crh-comment-Pull c3b565178724660cea8671850c8d12819a7ba4ef VersionControlUI/src/items/VDiffComparisonPair.cpp 113'></a> &nbsp;<b><a href='https://github.com/dimitar-asenov/Envision/pull/395#discussion_r66704029'>File: VersionControlUI/src/items/VDiffComparisonPair.cpp:L200-213</a></b>
- <a href='https://github.com/dimitar-asenov'></a> no clone on the two lines below
- [ ] <a href='#crh-comment-Pull c3b565178724660cea8671850c8d12819a7ba4ef VersionControlUI/src/nodes/DiffComparisonPair.h 18'></a> &nbsp;<b><a href='https://github.com/dimitar-asenov/Envision/pull/395#discussion_r66704064'>File: VersionControlUI/src/nodes/DiffComparisonPair.h:L68-77</a></b>
- <a href='https://github.com/dimitar-asenov'></a> You are storing pointers here and you did not explicitly delete those in the destructor so there is a memory leak.
  Unless you have a very specific reason why you need pointers (I'd like hear it if so), store directly the structures here and avoid the hassle of dealing with memory.
- [ ] <a href='#crh-comment-Pull c3b565178724660cea8671850c8d12819a7ba4ef VersionControlUI/src/nodes/DiffComparisonPair.cpp 87'></a> &nbsp;<b><a href='https://github.com/dimitar-asenov/Envision/pull/395#discussion_r66704101'>File: VersionControlUI/src/nodes/DiffComparisonPair.cpp:L155-176</a></b>
- <a href='https://github.com/dimitar-asenov'></a> I'm a bit confused. Now that we have the bread crumbs, do we still need this component name? Maybe I just forgot what it is.
- <a href='https://github.com/mgalbier'></a> It is used for the sorting of the DiffComparisonPairs. But there are fields which are no longer needed. I remove them as well as the " > " at the beginning of componentName, which does not serve any purpose.

<a href='https://www.codereviewhub.com/dimitar-asenov/Envision/pull/395?mark_as_completed=1'></a>&nbsp;<a href='https://www.codereviewhub.com/dimitar-asenov/Envision/pull/395?approve=1'></a>&nbsp;<a href='https://github.com/dimitar-asenov/Envision/pull/395'></a>
<a href='#crh-end'></a>

### CODE REVIEW

**Comment by dimitar-asenov** [comment]
This pull request is now managed with CodeReviewHub. <a href='https://github.com/dimitar-asenov/Envision/pull/395'>Refresh</a>.<a href='#crh-update-none'></a>

File: VersionControlUI/src/DiffManager.cpp [hunk 1]
```diff
@@ -276,16 +276,19 @@ bool DiffManager::findChangedNode(Model::TreeManager* treeManager, Model::NodeId
...
+	diffViewItem->setArrowStyle(modifyArrowLayer, "modify_arrow");
 
 	QSet<Visualization::Item*> allItemsToScale;
 	for (auto change : changesWithNodes)
 	{
 		QString highlightOverlayStyle;
 		QString highlightOverlayName;
-		QString arrowIconOverlayStyle;
-		QString arrowIconOverlayName;
+		QString arrowIconOverlayStyle = "";
```
Comment (dimitar-asenov) [inline VersionControlUI/src/DiffManager.cpp]: instead of assigning "" to the strings here, just leave them uninitialized. Below you can use `isNull()` to check if the strings have changed.

File: VersionControlUI/src/DiffManager.cpp [hunk 2]
```diff
@@ -381,9 +385,12 @@ Visualization::Item* DiffManager::addOverlaysAndReturnItem(Model::Node* node, Vi
 					Visualization::HighlightOverlay::itemStyles().get(highlightOverlayStyle)};
 			resultItem->addOverlay(overlay, highlightOverlayName);
 
-			auto iconOverlay = new Visualization::IconOverlay{resultItem,
-					Visualization::IconOverlay::itemStyles().get(arrowIconOverlayStyle)};
-			resultItem->addOverlay(iconOverlay, arrowIconOverlayName);
+			if (arrowIconOverlayName != "")
```
Comment (dimitar-asenov) [inline VersionControlUI/src/DiffManager.cpp]: use `isNull()`

File: VersionControlUI/src/DiffManager.cpp [hunk 3]
```diff
@@ -331,11 +330,16 @@ void DiffManager::createOverlaysForChanges(Visualization::ViewItem* diffViewItem
 		if (newNodeItem)
 			allItemsToScale.insert(newNodeItem);
 
-		if (change.changeType_== FilePersistence::ChangeType::Move)
+		if (change.changeType_== FilePersistence::ChangeType::Move && newNodeItem && oldNodeItem)
```
Comment (dimitar-asenov) [inline VersionControlUI/src/DiffManager.cpp]: why did you change this condition?

**Comment by dimitar-asenov** [comment]
I wouldn't delete the svgs and the styles corresponding to the yellow and blue arrows. Just leave them unused. Who knows if we change our minds in the future.

File: VersionControlUI/src/items/VDiffComparisonPair.cpp [hunk 4]
```diff
@@ -96,14 +117,23 @@ void VDiffComparisonPair::initializeForms()
 			->setColumnHorizontalAlignment(1, Visualization::LayoutStyle::Alignment::Left)
 			->setRowStretchFactor(0, 1)
 			->setRowStretchFactor(1, 1)
-			->put(0, 0, infoGrid->clone())
+			->put(0, 0, infoGrid)
```
Comment (dimitar-asenov) [inline VersionControlUI/src/items/VDiffComparisonPair.cpp:120]: In the line below and the `addForm` line you are unnecessarily cloning and thereby leaking memory. Don't clone anything, since you overwrite the `infoGrid` and `diffGrid` and `container` variables with new objects anyway later.

File: VersionControlUI/src/items/VDiffComparisonPair.cpp [hunk 5]
```diff
@@ -130,13 +160,22 @@ void VDiffComparisonPair::initializeForms()
 			->setRowStretchFactor(0, 1)
 			->setRowStretchFactor(1, 1)
 			->setColumnHorizontalAlignment(0, Visualization::LayoutStyle::Alignment::Center)
-			->put(0, 0, infoGrid->clone())
+			->put(0, 0, infoGrid)
```
Comment (dimitar-asenov) [inline VersionControlUI/src/items/VDiffComparisonPair.cpp:163]: no clone on next line and on the `addForm` line following

File: VersionControlUI/src/items/VDiffComparisonPair.cpp [hunk 6]
```diff
@@ -161,21 +200,13 @@ void VDiffComparisonPair::initializeForms()
 			->setRightMargin(10)
 			->setColumnStretchFactor(0, 1)
 			->setColumnHorizontalAlignment(0, Visualization::LayoutStyle::Alignment::Center)
-			->put(0, 0, infoGrid->clone())
+			->put(0, 0, infoGrid)
```
Comment (dimitar-asenov) [inline VersionControlUI/src/items/VDiffComparisonPair.cpp:203]: no clone on the two lines below

File: VersionControlUI/src/nodes/DiffComparisonPair.h [hunk 7]
```diff
@@ -61,6 +68,9 @@ class VERSIONCONTROLUI_API DiffComparisonPair : public Super<Visualization::UINo
 		bool twoObjectPathsDefined();
 		QString comparisonName();
 
+		QList<ObjectPathCrumbData*> objectPathCrumbsDataOldNode();
```
Comment (dimitar-asenov) [inline VersionControlUI/src/nodes/DiffComparisonPair.h]: You are storing pointers here and you did not explicitly delete those in the destructor so there is a memory leak.

Unless you have a very specific reason why you need pointers (I'd like hear it if so), store directly the structures here and avoid the hassle of dealing with memory.

File: VersionControlUI/src/nodes/DiffComparisonPair.cpp [hunk 8]
```diff
@@ -123,18 +155,21 @@ QString DiffComparisonPair::computeComponentName()
...
+	}
 
 	if (searchCompositeNode)
 	{
 		auto compositeNode = DCast<Model::CompositeNode>(searchCompositeNode);
 		if (compositeNode != nodeToComputeComponentType)
 		{
 			auto index = compositeNode->indexOf(nodeToComputeComponentType);
-			componentName = compositeNode->meta().attribute(index).name();
+			componentName = " > " + compositeNode->meta().attribute(index).name();
```
Comment (dimitar-asenov) [inline VersionControlUI/src/nodes/DiffComparisonPair.cpp]: I'm a bit confused. Now that we have the bread crumbs, do we still need this component name? Maybe I just forgot what it is.

**Comment by dimitar-asenov** [comment]
OK I'm done with this review.

File: VersionControlUI/src/DiffManager.cpp (hunk 3 above)
Comment (mgalbier) [inline VersionControlUI/src/DiffManager.cpp]: If one of the items is null it does not make sense to add an arrow between them. The if structure was changed in a later commit a bit. An outer if checks if the items are not null and only then checks the other stuff.

Comment (dimitar-asenov) [inline VersionControlUI/src/DiffManager.cpp]: OK, thanks for explaining.

File: VersionControlUI/src/nodes/DiffComparisonPair.cpp (hunk 8 above)
Comment (mgalbier) [inline VersionControlUI/src/nodes/DiffComparisonPair.cpp]: It is used for the sorting of the DiffComparisonPairs. But there are fields which are no longer needed. I remove them as well as the " > " at the beginning of componentName, which does not serve any purpose.

**Comment by mgalbier** [comment]
Thanks. Updated

**Comment by dimitar-asenov** [comment]
Nice, thanks.
